Better handling of HTTP error 503

dmitry-yudakov opened the pull request:
Redirecting to maintenance page is pretty bad user experience - especially after momentum has been load and was working successfully. You cannot simply reload the page because of changed url and the existing state potentially working gets lost.

That's why I made a change - if the first request (`config/ui-client`) fails with 503 or something else, then we show Maintenance or SmtWentWrong error without changing url.

Further API requests with 503 are retried up to 3 times with incremental delay (1 sec, 2 sec, 4 sec) - configurable. If it still fails after retry, toast notification is shown.

It's implemented without using external libraries as they didn't play too well with our existing logic.

dmitry-yudakov committed:
feat(API): exponential retry of HTTP error 503

Currently we redirect to maintenance page when a single error 503 is received - better retry it several times between failing it

PAKovalev commented on packages/app/src/scenes/App.tsx:
```diff
@@ -34,10 +34,20 @@ const App: FC = () => {
   const {t} = useTranslation();
 
   useEffect(() => {
+    setAxiosRetry({
```
Lets create a hook instead of useEffect

dmitry-yudakov committed:
Merge branch 'develop' into feat/axios-retry-maintenance-error

dmitry-yudakov committed:
feat: show maintenance screen only after 503 for config/ui-client, show toast otherwise, retry 503 errors

I removed axios-retry as it didn't play well with our existing onError logic.

I implemented similar logic compatible with handling errors - currently error codes 503 are retried up to 3 times and then onError logic is called.

No redirect to maintenance page is done as it's bad user experience after all - maintenance page message is shown if config/ui-client has failed with 503. In other cases 503 error is retried and if unsuccessful, toast is shown.

dmitry-yudakov committed:
fix(config-store): warning about any

dmitry-yudakov committed:
refactor: remove obsolete log

PAKovalev commented on packages/app/src/api/request/request.ts:
```diff
@@ -53,8 +53,48 @@ const getAccessToken = (): string => {
+  onError = errorHandler,
+  maxRetries = 3,
+  retryDelayBase = 1000,
+  retryCodes = [503]
```
Use existing constants.

PAKovalev commented on packages/app/src/api/request/request.ts:
```diff
@@ -53,8 +53,48 @@ const getAccessToken = (): string => {
+export const setApiResponseHandlers = ({
+  onResponse = responseHandler,
+  onError = errorHandler,
+  maxRetries = 3,
```
Let's move this to the env file.

dmitry-yudakov replied:
Similar parameters like response timeout are in this file, so I put the new ones next to them.

PAKovalev commented on packages/app/src/scenes/App.tsx:
```diff
@@ -33,24 +33,39 @@ import 'react-toastify/dist/ReactToastify.css';
   const history = useHistory();
   const {t} = useTranslation();
 
   useEffect(() => {
```
Lets create a hook under `shared/hooks`.

dmitry-yudakov replied:
I'll leave it as is for the moment. Lets discuss when you're back. I don't think it's good to hide all the logic of this effect in a hook, yet if it's just to spare useEffect+setApiResponseHandlers it's meaningless and only gives you pain of setting up the types and figuring out how to solve the issue of the inner deps list of the same useEffect.

Daboo1998 commented on packages/app/src/api/request/request.ts:
```diff
@@ -34,7 +34,7 @@ const responseHandler = (response: AxiosResponse) => {
 
 const errorHandler = (error: AxiosError) => {
-  throw error;
+  return Promise.reject(error) as unknown;
```
why not change to async and then throw error?

dmitry-yudakov replied:
It was needed for the first implementation, but now I just restored it to the original `throw error`.

Daboo1998 commented on packages/app/src/stores/ConfigStore/ConfigStore.ts:
```diff
@@ -25,7 +25,7 @@ const ConfigStore = types
         Object.entries(appVariablesOverrides).forEach((entry) => {
           const [key, value] = entry;
-          appVariables[key as keyof AppConfigInterface] = value as any;
+          appVariables[key as keyof AppConfigInterface] = value as string;
```
actually this is wrong. It should be 
```suggestion
          appVariables[key as keyof AppConfigExtendedInterface] = value as any;
```

as appVariables are of type `AppConfigExtendedInterface`

dmitry-yudakov replied:
It doesn't really matter - it's just casting of stuff and doesn't affect anything. And `any` will give us warning.

dmitry-yudakov replied:
And it gives error `Type 'any' is not assignable to type 'never'.` if AppConfigExtendedInterface is used, whatever the hell it means. Lets just leave it as is ;)

dmitry-yudakov committed:
refactor(api): move values to constants, simplify default error handler
